File: src/main/java/com/snooper/Application.java
package com.snooper;

import com.snooper.constant.Action;
import com.snooper.dto.Plateau;
import com.snooper.dto.RoboticRover;
import com.snooper.exception.OutOfBoundaryException;
import com.snooper.service.CommandHandler;
import com.snooper.service.CommandHandlerImpl;
import com.snooper.service.PlateauNavigation;
import com.snooper.service.PlateauNavigationImpl;

import java.util.ArrayList;
import java.util.List;

public class Application {

    PlateauNavigation plateauNavigation = new PlateauNavigationImpl();
    CommandHandler commandHandler = new CommandHandlerImpl();

    public String roboticRoverNavigatePlateau(String input) {
        String[] commands = input.split("\n");
        if (commands.length == 0) {
            throw new IllegalArgumentException("Null input");
        }

        if (commands.length % 2 == 0) {
            throw new IllegalArgumentException("Invalid input");
        }

        List<String> results = new ArrayList<>();
        Plateau plateau = commandHandler.parsePlateau(commands[0]);

        for (int i = 1; i < commands.length; i += 2) {
            RoboticRover roboticRover = commandHandler.parseRoboticRover(commands[i]);
            List<Action> actions = commandHandler.parseActions(commands[i + 1]);

            try {
                plateauNavigation.handle(plateau, roboticRover, actions);
                int x = roboticRover.getCoordinate().getX();
                int y = roboticRover.getCoordinate().getY();
                String direction = roboticRover.getDirection().toString();
                results.add(x + " " + y + " " + direction);
            } catch (OutOfBoundaryException e) {
                results.add(e.getMessage());
            }
        }

        StringBuilder stringBuilder = new StringBuilder();
        for (String result: results) {
            stringBuilder.append(result + "\n");
        }
        return stringBuilder.toString();
    }


    public static void main(String[] args) {
        Application application = new Application();
        System.out.println(application.roboticRoverNavigatePlateau(
                "5 5\n"
                + "1 2 N\n"
                + "LMLMLMLMM\n"
                + "3 3 E\n"
                + "MMRMMRMRRM"
        ));
    }
}

File: src/main/java/com/snooper/Application.kt
package com.snooper

import com.snooper.exception.OutOfBoundaryException
import com.snooper.service.CommandHandler
import com.snooper.service.CommandHandlerImpl
import com.snooper.service.PlateauNavigation
import com.snooper.service.PlateauNavigationImpl
import java.util.*
import java.util.stream.Collectors

class Application {

    private var plateauNavigation: PlateauNavigation = PlateauNavigationImpl()
    private var commandHandler: CommandHandler = CommandHandlerImpl()

    fun roboticRoverNavigatePlateau(input: String): String {
        val commands = input.split("\n")

        require(commands.isNotEmpty()) { "Null input" }
        require(commands.size % 2 != 0) { "Invalid input" }

        val plateau = commandHandler.parsePlateau(commands[0])

        // Each command set includes roboticRover and actions
        val commandSets: MutableList<Array<String>> = ArrayList()

        var i = 1
        while (i < commands.size) {
            commandSets.add(arrayOf(commands[i], commands[i + 1]))
            i += 2
        }

        // parallel running commandSets
        val results = commandSets.stream()
                .parallel()
                .map { commandSet: Array<String> ->
                    val roboticRover = commandHandler.parseRoboticRover(commandSet[0])
                    val actions = commandHandler.parseActions(commandSet[1])
                    try {
                        plateauNavigation.handle(plateau, roboticRover, actions)
                        val x = roboticRover.coordinate.x
                        val y = roboticRover.coordinate.y
                        val direction = roboticRover.direction.toString()
                        return@map "$x $y $direction"
                    } catch (e: OutOfBoundaryException) {
                        return@map e.message
                    }
                }
                .collect(Collectors.toList())
        val stringBuilder = StringBuilder()
        for (result in results) {
            stringBuilder.append(result + "\n")
        }
        return stringBuilder.toString()
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val application = Application()
            println(application.roboticRoverNavigatePlateau(
                    """
                        5 5
                        1 2 N
                        LMLMLMLMM
                        3 3 E
                        MMRMMRMRRM
                        0 0 S
                        M
                        1 1 W
                        M
                        2 2 N
                        M
                        """.trimIndent()
            ))
        }
    }
}